compass/_scrapers/reports.py

```diff
@@ -0,0 +1,149 @@
+            "ReportViewer1$ctl04$ctl05$divDropDown$ctl01$HiddenIndices": "0",
+        }
+
+        # districts = tree.xpath("//div[@id='ReportViewer1_ctl04_ctl07_divDropDown']//label/text()")
+        # list_districts = [unicodedata.normalize("NFKD", d) for d in districts if "select all" not in d.lower()]
+        # numbered_districts = {str(i): d for i, d in enumerate(list_districts)}
+        # all_districts = ", ".join(numbered_districts.values())
+        # all_districts_indices = ",".join(numbered_districts.keys())
+        #
+        # counties = tree.xpath("//div[@id='ReportViewer1_ctl04_ctl05_divDropDown']//label/text()")
+        # list_counties = [unicodedata.normalize("NFKD", c) for c in counties if "select all" not in c.lower()]
+        # numbered_counties = {str(i): c for i, c in enumerate(list_counties)}
+        # all_counties = ", ".join(numbered_counties.values())
+        # all_counties_indices = ",".join(numbered_counties.keys())
+        #
+        # form_data = {
+        #     "__VIEWSTATE": elements["__VIEWSTATE"],
+        #     "ReportViewer1$ctl04$ctl05$txtValue": all_counties,
+        #     "ReportViewer1$ctl04$ctl05$divDropDown$ctl01$HiddenIndices": all_counties_indices,
+        #     "ReportViewer1$ctl04$ctl07$txtValue": all_districts,
+        #     "ReportViewer1$ctl04$ctl07$divDropDown$ctl01$HiddenIndices": all_districts_indices,
+        # }
+
+        # Including MicrosoftAJAX: Delta=true reduces size by ~1kb but increases time by 0.01s.
+        # In reality we don't care about the output of this POST, just that it doesn't fail
+        report = self._post(run_report, data=form_data, headers={"X-MicrosoftAjax": "Delta=true"})
+        report.raise_for_status()
+
+        # Check error state
+        if "compass.scouts.org.uk%2fError.aspx|" in report.text:
+            raise CompassReportError("Compass Error!")
+
+    def report_keep_alive(self, report_page: str):
+        print(f"Extending Report Session {datetime.datetime.now()}")
+        keep_alive = re.search(r'"KeepAliveUrl":"(.*?)"', report_page).group(1).encode().decode("unicode-escape")
+        response = self._post(f"{Settings.base_url}{keep_alive}")  # NoQA: F841
+
+        return keep_alive  # response
+
+    def download_report_streaming(self, url: str, params: dict, filename: str):
+        with self._get(url, params=params, stream=True) as r:
+            print("")
+            r.raise_for_status()
+            with open(filename, "wb") as f:
+                for chunk in r.iter_content(chunk_size=1024 ** 2):  # Chunk size == 1MiB
+                    f.write(chunk)
+
+    def download_report_normal(self, url: str, params: dict, filename: str) -> bytes:
+        start = time.time()
+        csv_export = self._get(url, params=params)
+        print(f"Exporting took {time.time() - start}s")
+        print("Saving report")
+        Path(filename).write_bytes(csv_export.content)  # TODO Debug check
```

Comment:
Might be worth checking if this path is valid/writeable

Reply:
✅